kernel/aster-nix/src/util/net/addr/unix.rs

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::{ffi::CStr, mem::offset_of};
+
+use super::family::CSocketAddrFamily;
+use crate::{net::socket::unix::UnixSocketAddr, prelude::*};
+
+const PATH_MAX_LEN: usize = 108;
+
+/// UNIX domain socket address.
+///
+/// See <https://www.man7.org/linux/man-pages/man7/unix.7.html>.
+#[repr(C)]
+#[derive(Debug, Clone, Copy, Pod)]
+pub(super) struct CSocketAddrUnix {
+    /// Address family (AF_UNIX).
+    sun_family: u16,
+    /// Pathname.
+    sun_path: [u8; PATH_MAX_LEN],
+}
+
+impl CSocketAddrUnix {
+    const PATH_OFFSET: usize = offset_of!(Self, sun_path);
+
+    const MIN_LEN: usize = Self::PATH_OFFSET;
+    const MAX_LEN: usize = size_of::<Self>();
+
+    /// Converts from a [`UnixSocketAddr`] by encoding the pathname.
+    ///
+    /// We don't actually create a [`CSocketAddrUnix`]. Instead, we create its byte representation
+    /// directly for ease of operation.
+    ///
+    /// # Panics
+    ///
+    /// This method will panic if the pathname in [`UnixSocketAddr`] is too long to be stored in
+    /// the `sun_path` field of `CUnixSocketAddr`.
+    pub(super) fn encode<R, F>(value: &UnixSocketAddr, f: F) -> R
```

Comment:
The same argument applies to another method `decode`. So let's discuss whether the two methods `encode` and `decode` should belong to `CSockAddrUnix`.

At first glance, the method seems to be unrelated to `CSockAddrUnix`, but I have documented why it is related:
```rust
    /// Converts from a [`UnixSocketAddr`] by encoding the pathname.
    ///
    /// We don't actually create a [`CSocketAddrUnix`]. Instead, we create its byte representation
    /// directly for ease of operation.
```

I previously tried to make `encode` return `CSockAddrUnix`, but it doesn't work well because the `CSockAddrUnix` is a variable length structure, so I have to return `(CSockAddrUnix, usize)` instead. But considering the null terminator, the returned size can be one byte longer than `size_of::<CSockAddrUnix>()`. That's really annoying, but generating the byte representation of `CSockAddrUnix` directly will make things much easier.

Although I don't think "This method appears to have little correlation with `CSocketAddrUnix`", removing this method is fine by me because it doesn't really produce a `CSocketAddrUnix`. We can rename `CSockUnixAddr::encode` to `unix::into_c_bytes_and` and `CSockUnixAddr::decode` to `unix::from_c_bytes`. Hopefully this will also sound good to you.

Reply:
> But considering the null terminator, the returned size can be one byte longer than size_of::<CSockAddrUnix>(). That's really annoying, but generating the byte representation of CSockAddrUnix directly will make things much easier.

Yes, it makes sense. So it's impossible to return `CSockAddrUnix`.

> We can rename CSockUnixAddr::encode to unix::into_c_bytes_and and CSockUnixAddr::decode to unix::from_c_bytes.

Renaming these two methods looks good to me.